src/files.ml

```diff
@@ -3,6 +3,211 @@
+        raised if the root of [m] was not previously set with [set_root]. *)
+    val get : Path.t -> t -> file_path
+
+    (** [iter fn m] calls function [fn] on every binding stored in [m]. *)
+    val iter : (Path.t -> file_path -> unit) -> t -> unit
+  end =
+  struct
+    type t = Node of file_path option * t StrMap.t
+
+    let empty = Node(None, StrMap.empty)
+
+    exception Already_mapped
+
+    let set_root path (Node(po, m)) =
+      match po with
+      | None    -> Node(Some(path), m)
+      | Some(_) -> raise Already_mapped
+
+    let rec singleton ks path =
+      match ks with
+      | []      -> Node(Some(path), StrMap.empty)
+      | k :: ks -> Node(None, StrMap.singleton k (singleton ks path))
+
+    let rec add ks path (Node(po, m)) =
+      match (po, ks) with
+      | (Some(_), []     ) -> raise Already_mapped
+      | (None   , []     ) -> Node(Some(path), m)
+      | (_      , k :: ks) ->
+          try Node(po, StrMap.add k (add ks path (StrMap.find k m)) m)
+          with Not_found -> Node(po, StrMap.add k (singleton ks path) m)
+
+    exception Root_not_set
+
+    let get ks (Node(po, _) as m) =
+      let rec get root ks (Node(po, m)) =
+        match ks with
+        | []      -> Option.get po root
+        | k :: ks ->
+            try get (Filename.concat root k) ks (StrMap.find k m)
+            with Not_found -> List.fold_left Filename.concat root (k :: ks)
+      in
+      match po with
+      | None       -> raise Root_not_set
+      | Some(root) -> get root ks m
+
+    let iter fn m =
+      let rec iter path (Node(po, m)) =
+        Option.iter (fun file -> fn path file) po;
+        StrMap.iter (fun k m -> iter (path @ [k]) m) m
+      in
+      iter [] m
+  end
+
+(** [lib_root] stores the result of the ["--lib-root"] flag when given. *)
+let lib_root : string option Pervasives.ref =
+  Pervasives.ref None
+
+(** [set_lib_root path] sets the library root to the given file [path]. If the
+    given [path] does not refer to a valid (existing) directory the program is
+    terminated with a graceful error message. *)
+let set_lib_root : string -> unit = fun path ->
+  try
+    let path = Filename.realpath path in
+    if not (Sys.is_directory path) then
+      exit_with "Invalid library root ([%s] is not a directory)." path;
+    match Pervasives.(!lib_root) with
+    | None    -> Pervasives.(lib_root := Some(path))
+    | Some(_) -> exit_with "The library root was already set."
+  with Sys_error(_) | Invalid_argument(_) ->
+    exit_with "Invalid library root (no such file or directory [%s])." path
+
+(** [default_lib_root ()] returns the default library root curently configured
+    for the system. It depends on the current Opam switch (if any), and it may
+    or may not correspond to an existing directory. *)
+let default_lib_root : unit -> string = fun _ ->
+  match run_process "opam var lib" with
+  | Some([l]) -> Filename.concat l "lambdapi/lib_root"
+  | _         -> "/usr/local/lib/lambdapi/lib_root"
```

Comment:
Isn't it a problem to depend on some external system like opam?

Reply:
Or to hard-code paths like "/usr/local/lib/lambdapi/lib_root" which is probably OS dependent?

Reply:
> Isn't it a problem to enforce the use of a system like opam? What if the user wants to experiment with some libraries he modified and compiled somewhere in its home directory, for instance, in some git repos?

Well, that's what the `--map` option is for!

In any case, enforcing the use of a proper tool (nothing like opam, which is crap) is great. It makes everything simpler and more uniform, and the user does not have to care about completely useless details. And also, everything is predictable.

Reply:
Note that the installation of our `lambdapi` binary also depends on opam! There is nothing I can do about that.

Reply:
That's an optional dependency indeed. I'd recommend a variation on set of rules dune uses, c.f. https://github.com/ocaml/dune/pull/1226 so for example:

```
    let opam_prefix = Env.get env "OPAM_SWITCH_PREFIX" in
```
can be used for opam2, and also you allow the user to override that using a config step.

Reply:
> Note that the installation of our `lambdapi` binary also depends on opam! There is nothing I can do about that.

What do you mean? It doesn't depend on OPAM AFAICS; for emacs files you could just add the proper `(install ...)` rules and `dune install` would take care of it.

Reply:
Well, what I meant is that everything is gonna be installed under `opam` since dune uses its data. And since `dune` is not yet able to give me access to the location of the `lib` directory at compile time I don't have many choices here. I would indeed prefer to pick the `lib` directory during a config step.

Reply:
> That's an optional dependency indeed. I'd recommend a variation on set of rules dune uses, c.f. [ocaml/dune#1226](https://github.com/ocaml/dune/pull/1226) so for example:
> 
> ```
>     let opam_prefix = Env.get env "OPAM_SWITCH_PREFIX" in
> ```
> 
> can be used for opam2, and also you allow the user to override that using a config step.

I considered that, but is the `lib` directory always fixed with respect to the prefix?

Reply:
For all practical purposes yes.

Reply:
> Or to hard-code paths like "/usr/local/lib/lambdapi/lib_root" which is probably OS dependent?

Well, that's a reasonable directory for POSIX systems. Any other systems can't even handle OCaml so I think we are good.

Anyway, that is not the most pressing kind of issue here. I'm trying to make things work on most systems that we are interested in, and I can't solve every problem in the world (especially not at once). So I'm making what I see as reasonable assumptions, and we will always be able to make things better later.